plugins/catalog-backend/src/processing/TaskPipeline.ts

```diff
@@ -72,50 +74,58 @@ export function startTaskPipeline<T>(options: Options<T>) {
-    if (stopped || loading || inFlightCount > lowWatermark) {
-      return;
-    }
+  // State is in an object so that it can be stably referenced from within
+  // callbacks below
+  const state = { inFlightCount: 0 };
+  const abortController = new AbortController();
+  const abortSignal = abortController.signal;
 
-    // Once we hit the low watermark we load in enough items to reach the high watermark
-    loading = true;
-    const loadCount = highWatermark - inFlightCount;
-    const loadedItems = await loadTasks(loadCount);
-    loading = false;
+  async function pipelineLoop() {
+    const barrier = createBarrier({
+      waitTimeoutMillis: pollingIntervalMs,
+      signal: abortSignal,
+    });
 
-    // We might not reach the high watermark here, in case there weren't enough items to load
-    inFlightCount += loadedItems.length;
-    loadedItems.forEach(item => {
-      processTask(item).finally(() => {
-        if (stopped) {
-          return;
+    while (!abortSignal.aborted) {
+      if (state.inFlightCount <= lowWatermark) {
+        const loadCount = highWatermark - state.inFlightCount;
+        const loadedItems = await Promise.resolve()
+          .then(() => loadTasks(loadCount))
+          .catch(() => {
+            // Silently swallow errors and go back to sleep to try again; we
+            // delegate to the loadTasks function itself to catch errors and log
+            // if it so desires
+            return [];
+          });
+        if (loadedItems.length && !abortSignal.aborted) {
+          state.inFlightCount += loadedItems.length;
+          for (const item of loadedItems) {
+            Promise.resolve()
+              .then(() => processTask(item))
+              .catch(() => {
+                // Silently swallow errors and go back to sleep to try again; we
+                // delegate to the processTask function itself to catch errors
+                // and log if it so desires
+              })
+              .finally(() => {
+                state.inFlightCount -= 1;
+                barrier.release();
+              });
+          }
         }
+      }
 
-        // For each item we complete we check if it's time to load more
-        inFlightCount -= 1;
-        maybeLoadMore();
-      });
-    });
-
-    // We might have processed some tasks while we where loading, so check if we can load more
-    if (loadedItems.length > 1) {
-      maybeLoadMore();
+      await barrier.wait();
     }
   }
 
-  // This interval makes sure that we load in new items if the loop runs
-  // dry because of the lack of available tasks. As long as there are
-  // enough items to process this will be a noop.
-  const intervalId = setInterval(() => {
-    maybeLoadMore();
-  }, pollingIntervalMs);
+  pipelineLoop().catch(_error => {
+    // This should be impossible, but if it did happen, it would signal a
+    // programming error inside the loop (errors should definitely be caught
+    // inside of it).
+  });
```

Comment:
We don't want this to fail silently do we though? Probably best to re-throw as an uncaught exception to crash the backend?

Reply:
We don't crash on those anymore do we? 🤔

Reply:
Alright, updated this with a rethrow, it'll make it log which is excellent either way.